parsers/MN.py

```diff
@@ -1,16 +1,83 @@
 #!/usr/bin/env python3
 
+import json
 from datetime import datetime
 from logging import Logger, getLogger
-from typing import Optional
+from typing import Any, Dict, Optional
 
-from bs4 import BeautifulSoup
+import arrow
 from pytz import timezone
-from requests import Session
+from requests import Response, Session
+
+from parsers.lib.exceptions import ParserException
 
-#   MAIN_WEBSITE = https://ndc.energy.mn/
 NDC_GENERATION = "https://disnews.energy.mn/test/convert.php"
-TZ = "Asia/Ulaanbaatar"  # UTC+8
+TZ = timezone("Asia/Ulaanbaatar")  # UTC+8
+
+# Query fields to web API fields
+JSON_QUERY_TO_SRC = {
+    "time": "date",
+    "consumptionMW": "syssum",
+    "solarMW": "sumnar",
+    "windMW": "sums",
+    "importMW": "energyimport",  # positive = import
+    "temperatureC": "t",  # current temperature
+}
+
+
+def parse_json(web_json: dict) -> Dict[str, Any]:
+    """
+    Parse the fetched JSON data to our query format according to JSON_QUERY_TO_SRC.
+    Example of expected JSON format present at URL:
+    {"date":"2023-06-27 18:00:00","syssum":"869.37","sumnar":42.34,"sums":119.79,"energyimport":"49.58","t":"17"}
+    """
+
+    # Validate first if keys in fetched dict match expected keys
+    if set(JSON_QUERY_TO_SRC.values()) != set(web_json.keys()):
+        raise ParserException(
+            parser="MN.py",
+            message=f"Fetched keys from source {web_json.keys()} do not match expected keys {JSON_QUERY_TO_SRC.values()}.",
+        )
+
+    if None in web_json.values():
+        raise ParserException(
+            parser="MN.py",
+            message=f"Fetched values contain null. Fetched data: {web_json}.",
+        )
+
+    # Then we can safely parse them
+    query_data = dict()
+    for query_key, src_key in JSON_QUERY_TO_SRC.items():
+
+        if query_key == "time":
+            # convert to datetime
+            parsed_time = arrow.get(web_json[src_key], "YYYY-MM-DD HH:mm:ss", tzinfo=TZ)
+            query_data[query_key] = parsed_time.datetime
+        else:
+            # or convert to float, might also be string
+            query_data[query_key] = float(web_json[src_key])
+
+    return query_data
+
+
+def query(session: Session) -> Dict[str, Any]:
+    """
+    Query the JSON endpoint and parse it.
+    """
+
+    target_response: Response = session.get(NDC_GENERATION)
+
+    if not target_response.ok:
+        raise ParserException(
+            parser="MN.py",
+            message=f"Data request did not succeed: {target_response.status_code}",
+        )
+
+    # Read as JSON
+    response_json = json.loads(target_response.text)
+    query_result = parse_json(response_json)
```

Comment:
If it is actually json you can save a bit of trouble and simplify this a bit by calling response.json() instead of response.text

Reply:
Done!